README.md
## @observablehq/prettier-react-tachyons

![Node CI](https://github.com/observablehq/prettier-react-tachyons/workflows/Node%20CI/badge.svg)

A [prettier](https://prettier.io/) custom [parser plugin](https://prettier.io/docs/en/api.html#custom-parser-api)
for users of React and tachyons. It adds an extra prettier transform for `className=`
attributes on JSX elements, that takes care of the following:

- Eliminating redundant classes
- Sorting classes in a predictable order

### How do you use it?

This is admittedly a little bleeding-edge: prettier supports custom parsers,
but not custom rules / reformatters, so this essentially augments prettier's
`babylon`-based parser for JavaScript. You can configure prettier to point to it
with the `parser` option. For instance, here's how we configure prettier in our
`.prettierrc`:

```json
{
  "parser": "./node_modules/@observablehq/prettier-react-tachyons"
}
```

### Example:

in:

    <div className='h1 h1 w2 pv2 bg-transparent red' />

out:

    <div className='w2 h1 pv2 red bg-transparent' />


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [2.0.0](https://github.com/observablehq/prettier-react-tachyons/compare/v1.1.0...v2.0.0) (2020-05-05)


### ⚠ BREAKING CHANGES

* Node minimum version increased to 10

### Features

* Update dependencies including migration to post version 7 Babel name changes. ([32f9e61](https://github.com/observablehq/prettier-react-tachyons/commit/32f9e61b946fcf83ffee4d1a83ba09178b672e5e))
* Update to latest Babel parser ([6d6a6b6](https://github.com/observablehq/prettier-react-tachyons/commit/6d6a6b694608c772e580f13605b40095be8ec19c))


### Bug Fixes

* sort flex-column after inline-flex, make classes more exact ([6c83262](https://github.com/observablehq/prettier-react-tachyons/commit/6c83262ca350d05a9d2c8f7f95ef08700adfc815))

<a name="1.1.0"></a>
# [1.1.0](https://github.com/observablehq/prettier-react-tachyons/compare/v1.0.1...v1.1.0) (2018-03-27)


### Features

* Support v-base vertical-alignment option ([bafc544](https://github.com/observablehq/prettier-react-tachyons/commit/bafc544))


parser.test.js
const prettier = require("prettier");
const parser = require("./parser");

test("reformatting examples", () => {
  expect(
    prettier.format(`<div className='h1 h1 w2 pv2 bg-transparent red' />`, {
      parser
    })
  ).toMatchSnapshot();
  expect(
    prettier.format(`<div className='unknown pv2 ph2 w-50 h2' />`, {
      parser
    })
  ).toMatchSnapshot();
  expect(
    prettier.format(
      `<div className='fl dib pa2 mr3 black-50 hover-black bg-white ba b--black-30 br2 pointer' />`,
      {
        parser
      }
    )
  ).toMatchSnapshot();
  expect(
    prettier.format(`<div className=' h2 ' />`, {
      parser
    })
  ).toMatchSnapshot();
  expect(
    prettier.format(`<div className='inline-flex flex-column' />`, {
      parser
    })
  ).toMatchSnapshot();
});


test("works with typescript", () => {
  expect(
    prettier.format(`
      interface Props {
        prop?: string;
      }

      class Component extends React.Component<Props> {
        ref = React.createRef<HTMLCanvasElement>();
        property: string;
        constructor(props: Props) {
          super(props);
        }
        render() {
          return <div className='h1 h1 w2 pv2 bg-transparent red' />;
        }
      }
    `, {
      parser
    })
  ).toMatchSnapshot();
});


parser.js
const shades = ["05", "10", "20", "30", "40", "50", "60", "70", "80", "90"];
const colors = [
  "black",
  "near-black",
  "dark-gray",
  "mid-gray",
  "gray",
  "silver",
  "light-silver",
  "moon-gray",
  "light-gray",
  "near-white",
  "white",
  "dark-red",
  "red",
  "light-red",
  "orange",
  "gold",
  "yellow",
  "light-yellow",
  "purple",
  "light-purple",
  "dark-pink",
  "hot-pink",
  "pink",
  "light-pink",
  "dark-green",
  "green",
  "light-green",
  "navy",
  "dark-blue",
  "blue",
  "light-blue",
  "lightest-blue",
  "washed-blue",
  "washed-green",
  "washed-yellow",
  "washed-red"
]
  .concat(shades.map(s => `black-${s}`))
  .concat(shades.map(s => `white-${s}`));

const directions = "hvatlbr".split("");

// display, position, dimension, measure, flex, padding, margin, negative margin, font
// font size, font weight, line height, whitespace, color, opacity, background,
// hover background, border, border color,
// border radius
let order = [
  "(dn|di|db|dib|dit|dt|dtc|flex|inline-flex)",
  "(static|relative|absolute|fixed)",
  "(top|right|bottom|left)-\\d+",

  "measure",
  "measure-wide",
  "measure-narrow",
  "indent",
  "small-caps",
  "truncate",

  "w-?\\d+",
  "mw\\d+",
  "h\\d+",
  "center",
  "v-(base|mid|top|btm)",
  "(content|items|self|justify|order)",
  ...["l", "r", "n"].map(d => `f${d}`),
  ...directions.map(d => `p${d}\\d`),
  ...directions.map(d => `m${d}\\d`),
  ...directions.map(d => `n${d}\\d`),
  "(code|sans-serif|serif)",
  "f\\d",
  "fw\\d",
  "lh-w+",
  "(ws-normal|nowrap|pre)",
  "t[lrcj]",
  "(underline|no-underline|strike|ttu|ttc|ttl)",
  `(${colors.join("|")})`,
  `(${colors.map(c => `hover-${c}`).join("|")})`,
  "o\\d+",
  `(${colors.map(c => `bg-${c}`).join("|")})`,
  `(${colors.map(c => `hover-bg-${c}`).join("|")})`,
  ...directions.map(d => `b${d}$`),
  `(${colors.map(c => `b--${c}`).join("|")})`,
  "br[\\d-]"
].reduce((memo, re) => {
  return memo
    .concat(
      ["ns", "m", "l"].map(mq => {
        return new RegExp(`^${re}-${mq}$`);
      })
    )
    .concat(new RegExp(`^${re}$`));
}, []);

function weight(c) {
  for (let i = 0; i < order.length; i++) {
    if (order[i].test(c)) return i;
  }
  return Infinity;
}

module.exports = function(text, { babel }) {
  const traverse = require("@babel/traverse").default;
  const ast = babel(text);

  traverse(ast, {
    JSXAttribute({ node }) {
      if (
        node.name &&
        node.name.name == "className" &&
        node.value.type == "StringLiteral"
      ) {
        node.value.extra.raw = `"${Array.from(
          new Set(node.value.value.split(/\s+/).filter(Boolean)),
          (c, i) => [c, weight(c), i]
        )
          .sort(
            (left, right) =>
              left[1] !== right[1] ? left[1] - right[1] : left[2] - right[2]
          )
          .map(c => c[0])
          .join(" ")}"`;
      }
    }
  });
  return ast;
};
